### PR TITLE
[wip] Switch to zexe

### PR DESCRIPTION
**Pull request by dignifiedquire**
- [x] Implement missing circuits and proofs
  - [x] ~~`batchpost`~~ - will be deleted
  - [x] ~~`beacon_post`~~ - will be deleted
  - [x] `piece_inclusion_proofs`
  - [x] `vdf`
  - [x] `vdf_post`
  - [x] `vdf_sloth`
  - [x] `Blake2sHasher`
  - [x] `Sha256Hasher` no porting for now, as no sha256 circuit is implemented in zexe
  - [x] ~~`circuit/apex_commitment`~~ - will be deleted, and revived in a new form in the future by @porcuquine 
  - [x] `circuit/xor`
  - [x] `circuit/bench`
- [x] Switch storage-proofs-zexe to storage-proofs
- [ ] Publish zexe as crate and use it
- [x] Make filecoin-proofs work and use zexe proofs
- [x] Make fil-proofs-tooling work and use zexe proofs
- [ ] Reenable pip lifecycle tests, once pedersen implementation is fast enough

### CODE REVIEW

File: filecoin-proofs/examples/drgporep-vanilla.rs [hunk 1]
```diff
@@ -188,9 +188,9 @@ fn main() {
         "pedersen" => {
             do_the_work::<PedersenHasher>(data_size, m, challenge_count);
         }
-        "sha256" => {
-            do_the_work::<Sha256Hasher>(data_size, m, challenge_count);
-        }
+        // "sha256" => {
+        //     do_the_work::<Sha256Hasher>(data_size, m, challenge_count);
+        // }
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply filecoin-proofs/examples/drgporep-vanilla.rs]: unclear, which is why I left it there

File: storage-proofs/src/circuit/por.rs [hunk 2]
```diff
@@ -608,7 +578,7 @@ mod tests {
             assert!(cs.is_satisfied(), "constraints not satisfied");
 
             assert_eq!(cs.num_inputs(), 2, "wrong number of inputs");
-            assert_eq!(cs.num_constraints(), 4124, "wrong number of constraints");
+            assert_eq!(cs.num_constraints(), 25739, "wrong number of constraints");
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/circuit/por.rs]: yeah, we need to fix that pedersen impl

[the remark this answers is not in the document]
Comment (arielgabizon) [reply storage-proofs/src/circuit/por.rs]: That is the num of constraints for what exactly? The inefficient ped should 2.5-3 more constraints, here it's more than 6x.

[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/circuit/por.rs]: I am afraid things are worse than that, on `master` using sapling-crypto we have (byte size of the input, number of contraints)

```
(64, 1376),   // 64 bytes
(96, 2751),   // 96 bytes
(128, 4126),  // 128 bytes
(160, 5501),  // 160 bytes
(512, 20626), // 512 bytes
```

and using zexe in this PR

```
(64, 8576),    // 64 bytes
(96, 17152),   // 96 bytes
(128, 25728),  // 128 bytes
(160, 34304),  // 160 bytes
(512, 128640), // 512 bytes
```

I am not sure where the differences come from, maybe this has to do with the fact that we use the Merkle-Damgard construction?

**Comment by dignifiedquire** [comment]
This work is postponed for now

**Pull request closed without merging**